In [1]:
import pandas as pd

In [2]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")
products = pd.read_csv("../data/raw/olist_products_dataset.csv")
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")
translation = pd.read_csv("../data/raw/product_category_name_translation.csv")

In [3]:
Datasets ={
    'Customers': customers,
    'Geolocation': geolocation,
    'Orders': orders,
    'Order Items': order_items,
    'Payments': payments,
    'Reviews': reviews,
    'Products': products,
    'Sellers': sellers,
    'Translation': translation
}


In [4]:
def missing_data_summary(dataset):
    missing = dataset.isnull().sum()
    missing_percentage = (missing/len(dataset))*100
    missing_data = pd.DataFrame({'Missing Values': missing, 'Percentage': missing_percentage})
    return missing_data


for name, dataset in Datasets.items():
    print(f"Missing Data Summary for {name}:")
    display(missing_data_summary(dataset))
    print("\n")


Missing Data Summary for Customers:


,Missing Values,Percentage
customer_id,0,0.0
customer_unique_id,0,0.0
customer_zip_code_prefix,0,0.0
customer_city,0,0.0
customer_state,0,0.0




Missing Data Summary for Geolocation:


,Missing Values,Percentage
geolocation_zip_code_prefix,0,0.0
geolocation_lat,0,0.0
geolocation_lng,0,0.0
geolocation_city,0,0.0
geolocation_state,0,0.0




Missing Data Summary for Orders:


,Missing Values,Percentage
order_id,0,0.000000
customer_id,0,0.000000
order_status,0,0.000000
order_purchase_timestamp,0,0.000000
order_approved_at,160,0.160899
order_delivered_carrier_date,1783,1.793023
order_delivered_customer_date,2965,2.981668
order_estimated_delivery_date,0,0.000000




Missing Data Summary for Order Items:


,Missing Values,Percentage
order_id,0,0.0
order_item_id,0,0.0
product_id,0,0.0
seller_id,0,0.0
shipping_limit_date,0,0.0
price,0,0.0
freight_value,0,0.0




Missing Data Summary for Payments:


,Missing Values,Percentage
order_id,0,0.0
payment_sequential,0,0.0
payment_type,0,0.0
payment_installments,0,0.0
payment_value,0,0.0




Missing Data Summary for Reviews:


,Missing Values,Percentage
review_id,0,0.000000
order_id,0,0.000000
review_score,0,0.000000
review_comment_title,87656,88.341530
review_comment_message,58247,58.702532
review_creation_date,0,0.000000
review_answer_timestamp,0,0.000000




Missing Data Summary for Products:


,Missing Values,Percentage
product_id,0,0.000000
product_category_name,610,1.851234
product_name_lenght,610,1.851234
product_description_lenght,610,1.851234
product_photos_qty,610,1.851234
product_weight_g,2,0.006070
product_length_cm,2,0.006070
product_height_cm,2,0.006070
product_width_cm,2,0.006070




Missing Data Summary for Sellers:


,Missing Values,Percentage
seller_id,0,0.0
seller_zip_code_prefix,0,0.0
seller_city,0,0.0
seller_state,0,0.0




Missing Data Summary for Translation:


,Missing Values,Percentage
product_category_name,0,0.0
product_category_name_english,0,0.0


## Business Description for the missing values in the Datasets 

### 1. Customers
 No missing values in any column

### 2. Geolocation:
 No missing values in any column

### 3. Orders:
Orders_approved_at (160), order_delivered_carrier_date (1783) , order_delivered_customer_date (2965). 
Now since the dates values are missing, we would have difficulty in analysing some time related stuff i.e the time it took to deliver and whether it had impact on the reviews because the delivery time is not known.
Since some values of order approval dates are missing we would not be able to correctly tell the patttern of sales.

**2965 Order_delivered_customer_date records are missing.**

Business Reason 1: The order was cancelled, so it was never delivered.
Business Reason 2: The order was still processing, shipped, or in transit when the dataset was collected, so no delivery date existed yet.
Should we drop these rows?
No. These orders are still valuable because they provide information about cancellations, delivery performance, and the order fulfillment process. We should only exclude them if our analysis specifically requires completed deliveries.

### 4. Order Items: 
 No missing values in any column

### 5. Payments:
 No missing values in any column

### 6. Reviews:
review_comment_title(87656), review_comment_message(5824). It shows that reviews are incomplete. We can tell if the customer is satisfied or not but we can't trcak down the reason.

### 7. Products:
product_category_name(610), product_name_lenght	(610),	product_description_lenght(610),	product_photos_qty(610)
product_weight_g(2), product_length_cm(2), product_height_cm(2), product_width_cm(2).
We cannot do the correct analysis on the basis of category like which category is generating highest revenue or lowest revenue. The product quantity is missing as well so we will get inconsistent analysis results for multiple queries involving it.

### 8. Sellers:
 No missing values in any column

### 9. Translation: 
 No missing values in any column


In [5]:
def duplicate_data_summary(dataset):
    duplicates = dataset.duplicated().sum()
    duplicate_percentage = (duplicates / len(dataset)) * 100

    return pd.DataFrame({
        "Metric": ["Duplicate Rows", "Duplicate Percentage"],
        "Value": [duplicates, round(duplicate_percentage, 2)]
    })

for name, dataset in Datasets.items():
    print(name)
    display(duplicate_data_summary(dataset))
    print("\n")


Customers


,Metric,Value
0,Duplicate Rows,0.0
1,Duplicate Percentage,0.0




Geolocation


,Metric,Value
0,Duplicate Rows,261831.00
1,Duplicate Percentage,26.18




Orders


,Metric,Value
0,Duplicate Rows,0.0
1,Duplicate Percentage,0.0




Order Items


,Metric,Value
0,Duplicate Rows,0.0
1,Duplicate Percentage,0.0




Payments


,Metric,Value
0,Duplicate Rows,0.0
1,Duplicate Percentage,0.0




Reviews


,Metric,Value
0,Duplicate Rows,0.0
1,Duplicate Percentage,0.0




Products


,Metric,Value
0,Duplicate Rows,0.0
1,Duplicate Percentage,0.0




Sellers


,Metric,Value
0,Duplicate Rows,0.0
1,Duplicate Percentage,0.0




Translation


,Metric,Value
0,Duplicate Rows,0.0
1,Duplicate Percentage,0.0


## Business Description: 
No dataset contains duplicate rows, except the geolocation(fan-out problem) which is solved in the end. 

In [6]:
orders_duplicate = orders['order_id'].duplicated().sum()
display(orders_duplicate)

np.int64(0)

In [7]:
print("Duplicate customer_ids:", customers['customer_id'].duplicated().sum())
customers['customer_unique_id'].value_counts().head(10)

Duplicate customer_ids: 0


customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
dc813062e0fc23409cd255f7f53c7074     6
63cfc61cee11cbe306bff5857d00bfe4     6
f0e310a6839dce9de1638e0fe5ab282a     6
de34b16117594161a6a89c50b289d35a     6
Name: count, dtype: int64

The first line helps us record the duplicate results which helps us identify the problematic stuff in the dataset.
most customers in this dataset appear only once. One customer appearing 17 times is exceptional, not typical. This means Olist has a very low repeat purchase rate, which is a major business problem. A healthy e-commerce platform wants customers coming back. If most customers buy once and disappear, that is a retention crisis.

The second line shows which unique customers appeared most of the times so helps us see our loyal customers.

In [8]:
date_cols = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

for col in date_cols:
    orders[col] = pd.to_datetime(orders[col])
    print(orders[col].dtype)

datetime64[us]
datetime64[us]
datetime64[us]
datetime64[us]
datetime64[us]


In [9]:
order_items['shipping_limit_date'] = pd.to_datetime(order_items['shipping_limit_date'])
print(order_items['shipping_limit_date'].dtype)

datetime64[us]


In [10]:
products = products.rename(columns={
    'product_name_lenght': 'product_name_length',
    'product_description_lenght': 'product_description_length'
})

display(products.columns.tolist())

['product_id',
 'product_category_name',
 'product_name_length',
 'product_description_length',
 'product_photos_qty',
 'product_weight_g',
 'product_length_cm',
 'product_height_cm',
 'product_width_cm']

In [11]:
print("Shape before removing duplicates:")
print(geolocation.shape)

Shape before removing duplicates:
(1000163, 5)


In [12]:
geolocation_clean = geolocation.drop_duplicates(
    subset= 'geolocation_zip_code_prefix'
)

print("Shape after removing duplicates:")
print(geolocation_clean.shape)

rows_removed = len(geolocation) - len(geolocation_clean)

print(f"Rows removed: {rows_removed}")

Shape after removing duplicates:
(19015, 5)
Rows removed: 981148


### Deduplicating the Geolocation Table

The geolocation table contains multiple latitude and longitude coordinates for the same ZIP code prefix. This is expected because a ZIP code prefix can cover many different addresses within the same area.

If we merge this table directly with customers or sellers using the ZIP code prefix, a single customer could match multiple geolocation rows. This causes a **fan-out problem**, where one original row expands into several rows after the merge.

Fan-out can inflate row counts and produce incorrect business metrics, such as overstated sales totals or duplicated customer records.

To prevent this, we keep only one representative coordinate for each ZIP code prefix using `drop_duplicates(subset='geolocation_zip_code_prefix')`. This ensures a one-to-one relationship between ZIP code prefixes and geolocation records, making future joins safe and preventing duplicate rows from being introduced.

# Data Quality Report

## Critical Issues Found

- Missing values in several tables, especially product attributes and delivery dates.
- Multiple geolocation records for the same ZIP code prefix, which could create fan-out problems during joins.
- Duplicate ZIP code prefixes in the geolocation dataset.
- Customer records use both `customer_id` and `customer_unique_id`, requiring careful interpretation during analysis.

---

## Issues Fixed in This Notebook

- Investigated missing values for all tables.
- Calculated missing value counts and percentages.
- Checked duplicate rows in every table.
- Removed duplicate ZIP code prefixes from the geolocation dataset using `drop_duplicates()`.
- Documented the business meaning of missing values instead of removing data without justification.

---

## Known Limitations

- Missing values were identified but not imputed because appropriate replacement strategies depend on future analyses.
- Only ZIP code prefix duplicates were removed; duplicate coordinates may still exist at finer geographic levels.
- Business logic for handling cancelled and undelivered orders has not yet been implemented.
- Outlier detection and data validation have not been performed yet.

---

## Recommended Next Steps

- Handle missing values based on the requirements of each analysis.
- Validate data consistency across related tables after merging.
- Investigate delivery delays, cancellations, and review scores.
- Create a master analytical dataset by joining the cleaned tables.
- Begin exploratory data analysis (EDA) to identify trends, customer behavior, and business insights.